pkg/resources/resources.go

```diff
@@ -327,3 +329,54 @@ func NamespacedNameFromObject(obj client.Object) types.NamespacedName {
 		Name:      obj.GetName(),
 	}
 }
+
+// RemoveOwnerReferences removes all owner references from a Kubernetes object that match the provided predicate.
+//
+// This function iterates through the OwnerReferences of the given object, filters out those that satisfy
+// the predicate, and updates the object in the cluster using the provided client.
+//
+// Parameters:
+//   - ctx: The context for the request, which can carry deadlines, cancellation signals, and other request-scoped values.
+//   - cli: A controller-runtime client used to update the Kubernetes object.
+//   - obj: The Kubernetes object whose OwnerReferences are to be filtered. It must implement client.Object.
+//   - predicate: A function that takes an OwnerReference and returns true if the reference should be removed.
+//
+// Returns:
+//   - An error if the update operation fails, otherwise nil.
+func RemoveOwnerReferences(
+	ctx context.Context,
+	cli client.Client,
+	obj client.Object,
+	predicate func(reference metav1.OwnerReference) bool,
+) error {
+	oldRefs := obj.GetOwnerReferences()
+	if len(oldRefs) == 0 {
+		return nil
+	}
+
+	newRefs := oldRefs[:0]
```

Comment:
Is it to avoid array copy if there are no removes in the middle?

Reply:
yep, it is about to use the same underlying array to avoid allocations